[DI - step1] 철시 미션 제출합니다.

pkch93 opened the pull request:
안녕하세요! 하비 :) 이번에 리뷰를 부탁드리게 된 철시입니다 ㅎㅎ

이번 미션 구현하면서 가진 의문사항이 있어서 이렇게 여쭙습니다.

- 재귀 함수 리펙토링 관련
initializeInjectedBean2 메서드가 리펙토링 전의 재귀함수인데, 현재 저희는 initializeInjectedBean() 메서드와 같이 메서드로 분리해서 리펙토링했습니다. 그런데 이와 같이 하면 재귀 함수를 따라가기가 힘들다고 느껴졌습니다. 혹시 initializeInjectedBean()와 같이 분리하는 것이 좋은지, initializeInjectedBean2() 처럼 그대로 두는 것이 좋은지 리뷰어님의 생각이 궁금합니다.

- assert 구문 관련
inject() 메서드에서 injectedConstructor 변수에 대한 NULL체크를 assert를 통해 하고 있습니다. 그런데 이전에 호출되는 함수인 initializeNotInjectedBeans() 메서드에서 이미 NULL체크를 하고 있습니다. 그러면 NULL체크를 두 번 해야 하는 상황인데, 이렇게 중복해서 체크하더라도 명시적으로 assert를 통해 체크하는 것이 좋은지 리뷰어님의 생각이 궁금합니다.

잘 부탁드리겠습니다. 감사합니다!

pkch93 committed:
docs: 요구사항 정리

pkch93 committed:
feat: BeanFactory initialize 메서드 구현

pkch93 committed:
refactor: BeanFactory.initialize() 메서드 리팩토링

pkch93 committed:
docs: 의문사항 작성

pkch93 committed:
refactor: typo 수정

pkch93 committed:
docs: todo 추가

pkch93 committed:
feat: BeanScanner 구현

pkch93 committed:
refactor: AnnotationHandlerMapping에 BeanFactory 주입

young891221 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
+        }
+    }
+
+    private void initializeInjectedBeans() {
```
질문주신 부분에 대해서는 현재 방식이 더 좋아보이네요. 그 이유로는 현재와 같이 굳이 재귀를 돌릴 필요가 없어 보이며 가독성 측면에서 훨씬 간결해 보이기 때문입니다.

young891221 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanScanner.java:
```diff
@@ -0,0 +1,32 @@
+    @SuppressWarnings("unchecked")
+    public static Set<Class<?>> scan(String basePackage) {
+        Reflections reflections = new Reflections(basePackage);
+        return getTypesAnnotatedWith(reflections, Controller.class, Service.class, Repository.class);
```
해당하는 어노테이션을 활용하는 예제 + 테스트를 추가로 작성해 주세요.

pkch93 replied:
추가로 다른 패키지에 있는 컴포넌트를 스캔하는지와 다른 어노테이션이 붙은 컴포넌트를 스캔하지 않는지 테스트용 클래스와 함께 추가하였습니다!

young891221 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactoryUtils.java:
```diff
@@ -17,6 +17,7 @@
+    // TODO: 19. 10. 31. @Inject가 여러개인 경우 처리 방법 고민
     @SuppressWarnings({"rawtypes", "unchecked"})
     public static Constructor<?> getInjectedConstructor(Class<?> clazz) {
         Set<Constructor> injectedConstructors = getAllConstructors(clazz, withAnnotation(Inject.class));
```
Inject annotation에 대한 테스트도 추가해 주세요.

pkch93 replied:
`@Inject`가 붙은 생성자가 여러개인경우 에러를 발생하도록 변경하였습니다. 이와 함께 `@Inject`가 없는 경우 null을 반환하는지와 `@Inject`가 붙은 생성자가 여러개인 경우 에러가 발생하는지 테스트를 추가했습니다.

young891221 commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,139 @@
 
+    private Object inject(Class concreteClass) {
+        Constructor<?> injectedConstructor = BeanFactoryUtils.getInjectedConstructor(concreteClass);
+        assert injectedConstructor != null;
```
null 체크에 대한 부분은 각 메서드 단위로 하는 방향이 옳다고 생각됩니다.
해당 메서드가 추후 다른 곳에서 사용될 수 있다면 메서드 단위의 밸리데이션 체크는 기본으로 들어가야 하기 때문에 중복으로 null을 체크해도 무방합니다.

young891221 reviewed (changes requested):
안녕하세요 철시!
기본적인 사항들을 잘 만들어 주셨네요 👍 
질문 주신 사항과 더불어 몇 가지 피드백 남겼습니다. 확인해 주세요~!

pkch93 committed:
refactor: BeanFactory inject, getInstance 메서드명 변경

pkch93 committed:
refactor: BeanFactory initialize 메서드 중복제거

- initializeNotInjectedBeans 메서드가 initializeInjectedBeans와 같은 기능을
수행하고 있으므로 중복 부분 제거

pkch93 committed:
refactor: test 코드 package 구조 변경

pkch93 committed:
refactor: BeanScannerTest의 scanAll 메서드 DisplayName 추가

pkch93 committed:
docs: 추가 테스트 목록 작성

pkch93 committed:
test: BeanScanner 테스트케이스 추가

pkch93 committed:
docs: 의문사항 추가

pkch93 committed:
refactor: BeanFactoryUtils getInjectedConstructor 메서드

1. 테스트 케이스를 추가하였습니다.
2. Injected 생성자가 여러 개일 경우 예외를 던지도록 수정하였습니다.

pkch93 committed:
refactor: BeanFactoryUtils getInjectedConstructor 메서드

1. 테스트 케이스를 추가하였습니다.
2. Injected 생성자가 여러 개일 경우 예외를 던지도록 수정하였습니다.

pkch93 committed:
refactor: di 프레임워크 적용하여 Controller, Dao 리펙토링

pkch93 committed:
refactor: AnnotationHandlerMapping

- addHandlerExecutions 메서드 불필요한 인자 제거
- BeanFactory.getControllers 메서드가 Set 반환하도록 변경

pkch93 committed:
refactor: BeanFactory.initialize 메서드에서 불필요한 bypass 제거

pkch93 committed:
test: BeanFactory

- 생성자 매개변수로 같은 종류의 어노테이션을 가진 빈을 생성하는 테스트 구현

pkch93 commented:
피드백 감사합니다!

주신 피드백 반영하면서 같이 미션을 진행했던 페어의 피드백도 함께 반영했습니다! 그 과정에서 코드에 몇가지 변경이 있었습니다.

1. 기존에 BeanFactory에서 initialize 할 때 `@Inject` 생성자가 없는 컴포넌트를 빈으로 만드는 과정을 거친 후 `@Inject` 생성자가 있는 컴포넌트를 빈으로 만드는 과정을 거쳤습니다.

이 과정을 다시 한번 토의했는데 결국은 `@Inject` 생성자가 있는 컴포넌트를 스캔할 때 모든 빈이 만들어질 것이라 생각하여 위 과정에 중복이 있음을 알 수 있었습니다. 따라서 `@Inject` 생성자가 없는 컴포넌트를 빈으로 만드는 과정을 제거하고 `@Inject` 생성자가 있는 컴포넌트로만 빈을 만들도록 수정하였습니다.

2. 테스트 코드를 작성하면서 테스트에 필요한 클래스가 있어서 테스트 패키지에 테스트를 위한 패키지와 클래스를 생성했습니다.

3. 저희가 구현한 di 프레임워크가 제대로 작동하는지 확인하기위해 slipp 패키지에 적용해보고 테스트해보았습니다.

피드백과 리펙토링을 진행하면서 테스트 코드를 작성하면서 테스트만을 위한 패키지를 만드는 것에 대해 의문이 있었는데 혹시 하비는 어떻게 생각하시나요? 테스트 패키지에 생성한다면 테스트를 위해서는 만들어도 괜찮을까요??

young891221 reviewed (approved):
피드백 반영 감사합니다. 철시~! 💯 
저같은 경우 테스트를 위한 페키지를 생성하기 보다는 기능을 성격별 페키지로 분리하여 페키지로 접근 가능한
테스트가 되도록 합니다. 실제로 페키지 접근 지정자를 사용한 클래스나 메서드의 경우 이러한 방식의 테스트를 작성하는게 가장 올바르다고 생각해서요ㅎㅎ
고생하셨습니다. 머지할게요~!
